File: programmers/js/new/level2/소수_만들기.js
let answer = 0;
function solution(nums) {
    const isPrime = Array(3000).fill(false);
    const checked = Array(3000).fill(false);
    for (let i = 2; i < 3000; ++i) {
        if (!checked[i]) {
            isPrime[i] = true;
            for (let j = i; j < 3000; j += i) {
                checked[j] = true;
            }
        }
    }
    const selected = Array(nums.length).fill(false);
    dfs(selected, 0, 0, nums, isPrime);
    return answer;
}

function dfs(selected, idx, count, nums, isPrime) {
    if (count === 3) {
        let sum = 0;
        for (let i = 0; i < selected.length; ++i) {
            if (selected[i]) {
                sum += nums[i];
            }
        }
        if (isPrime[sum]) {
            answer++;
        }
        return;
    }
    if (idx === selected.length) {
        return;
    }
    selected[idx] = true;
    dfs(selected, idx + 1, count + 1, nums, isPrime);
    selected[idx] = false;
    dfs(selected, idx + 1, count, nums, isPrime);
}

solution([1, 2, 3, 4]);

File: programmers/python/level1/소수_만들기.py
answer = 0
is_prime = [False for i in range(3001)]
def solution(nums):
    checked = [False for i in range(3001)]
    for i in range(2,3001):
        if checked[i] == False:
            is_prime[i] = True
            for j in range(i, 3001, i):
                checked[j] = True
    visited = [False for i in range(len(nums))]
    dfs(0, visited, 0, nums)
    return answer

def dfs(count, visited, idx, nums):
    if count == 3:
        s = 0
        for i in range(len(nums)):
            if visited[i] == True:
                s += nums[i]
        if is_prime[s]:
            global answer
            answer += 1
        return
    if idx == len(nums):
        return
    visited[idx] = True
    dfs(count+1, visited, idx+1, nums)
    visited[idx] = False
    dfs(count, visited, idx+1, nums)

solution([1,2,3,4])